src/util/mod.rs

```diff
@@ -0,0 +1,149 @@
+use rand::prelude::*;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+const DEFAULT_BACKOFF_TIME_MAX_MS: f64 = 24f64 * 60f64 * 60f64 * 1000f64;
+const DEFAULT_BACKOFF_POW: f64 = 1.05f64;
+const DEFAULT_BACKOFF_TIME_MS: f64 = 5f64 * 1000f64;
+
+#[napi]
+
+pub struct QueueWithBackoff {
+    queue: VecDeque<String>,
+    /**
+     * A map of absolute backoff timestamps mapped to the value.
+     */
+    backoff: BTreeMap<u64, String>,
+    /**
+     * The last duration applied when a value was backed off.
+     */
+    last_backoff_duration: HashMap<String, u32>,
+    backoff_time: f64,
+    backoff_exponent: f64,
+    backoff_max: f64,
+}
+
+impl Default for QueueWithBackoff {
+    fn default() -> Self {
+        Self::new(
+            DEFAULT_BACKOFF_TIME_MS,
+            DEFAULT_BACKOFF_POW,
+            DEFAULT_BACKOFF_TIME_MAX_MS,
+        )
+    }
+}
+#[napi]
+
+impl QueueWithBackoff {
+    #[napi(constructor)]
+    pub fn new(backoff_time: f64, backoff_exponent: f64, backoff_max: f64) -> Self {
+        QueueWithBackoff {
+            queue: VecDeque::new(),
+            backoff: BTreeMap::new(),
+            last_backoff_duration: HashMap::new(),
+            backoff_time,
+            backoff_exponent,
+            backoff_max,
+        }
+    }
+
+    #[napi]
+    pub fn pop(&mut self) -> Option<String> {
+        let start = SystemTime::now();
+        let since_the_epoch = start.duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
+
+        // We only need to check this once, as we won't be adding to the backoff queue
+        // as often as we pull from it.
+        if let Some(item) = self.backoff.first_entry() {
+            if *item.key() < since_the_epoch {
+                let v = item.remove();
+                self.queue.push_back(v);
+            }
+        }
+
+        self.queue.pop_front()
+    }
+
+    #[napi]
+    pub fn remove(&mut self, item: String) -> bool {
+        // Remove from the queue
+        if let Ok(index) = self.queue.binary_search(&item) {
+            self.queue.remove(index);
+            return true;
+        } else {
+            // We didn't find the key queued, so let's ensure we delete it
+            // from any backoff.
+            // This is *expensive* but hopefully called rarely.
+            let mut found_key: u64 = 0;
+            for (key, value) in self.backoff.iter() {
+                if *value == item {
+                    found_key = *key;
+                }
+            }
+            if found_key != 0 {
+                self.backoff.remove(&found_key);
+                return true;
+            }
+        }
+        // Always remove the duration on removal.
+        self.last_backoff_duration.remove(&item);
```

Comment:
When would an item without a backoff have an entry in this?
Also, should this removal be done even when the entry does have a backoff?

Reply:
So if the item has never backed off, this is empty. If a entry that had previously failed had started working again, then we remove it. This map is purely to track that if we repeatedly back off an item, that we actually do increase the timeout rather than just timing out for the same duration each time.